// @tanstack/vue-query@5.90.5/build/modern/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  MutationCache: () => import_mutationCache.MutationCache,
  QueryCache: () => import_queryCache.QueryCache,
  QueryClient: () => import_queryClient.QueryClient,
  VUE_QUERY_CLIENT: () => import_utils.VUE_QUERY_CLIENT,
  VueQueryPlugin: () => import_vueQueryPlugin.VueQueryPlugin,
  infiniteQueryOptions: () => import_infiniteQueryOptions.infiniteQueryOptions,
  queryOptions: () => import_queryOptions.queryOptions,
  useInfiniteQuery: () => import_useInfiniteQuery.useInfiniteQuery,
  useIsFetching: () => import_useIsFetching.useIsFetching,
  useIsMutating: () => import_useMutationState.useIsMutating,
  useMutation: () => import_useMutation.useMutation,
  useMutationState: () => import_useMutationState.useMutationState,
  useQueries: () => import_useQueries.useQueries,
  useQuery: () => import_useQuery.useQuery,
  useQueryClient: () => import_useQueryClient.useQueryClient
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("@tanstack/query-core"), module.exports);
var import_useQueryClient = require("./useQueryClient.cjs");
var import_vueQueryPlugin = require("./vueQueryPlugin.cjs");
var import_queryClient = require("./queryClient.cjs");
var import_queryCache = require("./queryCache.cjs");
var import_queryOptions = require("./queryOptions.cjs");
var import_infiniteQueryOptions = require("./infiniteQueryOptions.cjs");
var import_mutationCache = require("./mutationCache.cjs");
var import_useQuery = require("./useQuery.cjs");
var import_useQueries = require("./useQueries.cjs");
var import_useInfiniteQuery = require("./useInfiniteQuery.cjs");
var import_useMutation = require("./useMutation.cjs");
var import_useIsFetching = require("./useIsFetching.cjs");
var import_useMutationState = require("./useMutationState.cjs");
var import_utils = require("./utils.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MutationCache,
  QueryCache,
  QueryClient,
  VUE_QUERY_CLIENT,
  VueQueryPlugin,
  infiniteQueryOptions,
  queryOptions,
  useInfiniteQuery,
  useIsFetching,
  useIsMutating,
  useMutation,
  useMutationState,
  useQueries,
  useQuery,
  useQueryClient,
  ...require("@tanstack/query-core")
});
